### PR TITLE
Be more resilient to broken deployments

### PR DESCRIPTION
**Pull request by LeFrosch**
If there is no `grpc-status` found in the header, I try to convert the http status into the corresponding grpc status. For the conversion I used the `fromHttpStatus` function from [this](https://github.com/grpc/grpc-web/blob/master/javascript/net/grpc/web/statuscode.js) as a template.
This fixes the issue I had here #421

### CODE REVIEW

File: lib/src/client/call.dart
Comment (mraleph) [inline lib/src/client/call.dart:357]: I think this can be `StatusCode.unknown` or `StatusCode.internal` - if response does not contain either `:status` or `grpc-status` then something is serious wrong with it.

File: lib/src/shared/status.dart
Comment (mraleph) [inline lib/src/shared/status.dart:127]: I suggest using a map here: 

```dart
const statusMapping = {200: Status.ok, 400: StatusCode.invalidArgument, ...};

return statusMapping[status] ?? StatusCode.unknown;
```

Even better for readability would be if you do:

```dart
// It is okay to import dart:io into shared code as long as you don't use
// any runtime functionality.
import 'dart:io' show HttpStatus;

const _httpStatusToGrpcStatus = <int, int>{
  HttpStatus.ok: StatusCode.ok,
  HttpStatus.badRequest: StatusCode.invalidArgument,
  HttpStatus.unauthorized: StatusCode.unauthenticated,
  HttpStatus.forbidden: StatusCode.permissionDenied,
  HttpStatus.notFound: StatusCode.notFound,
  HttpStatus.conflict: StatusCode.aborted,
  HttpStatus.preconditionFailed: StatusCode.failedPrecondition,
  HttpStatus.tooManyRequests: StatusCode.resourceExhausted,
  HttpStatus.clientClosedRequest: StatusCode.cancelled,
  HttpStatus.internalServerError: StatusCode.unknown,
  HttpStatus.notImplemented: StatusCode.unimplemented,
  HttpStatus.serviceUnavailable: StatusCode.unavailable,
  HttpStatus.gatewayTimeout: StatusCode.deadlineExceeded,
};
```

**Review by mraleph — CHANGES_REQUESTED** [review]
This looks good. Thank you for taking time to make a PR. 

I have made few suggestions how to make it a bit more robust. 

Please consider also adding at least few tests to make sure we don't regress this, but it's also okay to skip that.

File: lib/src/client/call.dart
Comment (LeFrosch) [reply lib/src/client/call.dart:354]: The first line doesn’t compile with null safety, because `metadata.containsKey(':status')` does not ensure that `metadata[':status']` cannot return null and `int.tryParse` expects a not nullable String. I’ll just tweak it a little bit to make it work with null safety.
But I noticed the header validation breaks a lot of tests, because most of them don’t populate the headers with the `status` and the `content-type`. In addition I tried the header validation with our asp .net core server and the server doesn’t seem to send the content type, thus this change would break our application. 
Personally I would suggest to only check if the http code is present when validating the header.

Comment (mraleph) [reply lib/src/client/call.dart:354]: > In addition I tried the header validation with our asp .net core server and the server doesn’t seem to send the content type, thus this change would break our application.

Interesting. That's a violation of gRPC specifications. I see that https://github.com/grpc/grpc-dotnet seems to send back correct Content-Type, are you using some custom gRPC implementation for .NET? I also see that most native clients validate Content-Type they receive.

Comment (mraleph) [reply lib/src/client/call.dart:354]: I also looked at grpc-web implementation and it does validate content type it receives from the server: https://github.com/grpc/grpc-web/blob/master/javascript/net/grpc/web/grpcwebclientreadablestream.js#L138

Comment (LeFrosch) [reply lib/src/client/call.dart:354]: My bad, it was because of something else that the content-type was not sent in my case. Sorry for wasting your time.

Comment (mraleph) [reply lib/src/client/call.dart:354]: No worries! You are *not* wasting my time in any way. Thank *you* for spending time on this PR, this is really appreciated.

File: test/client_tests/client_test.dart
Comment (LeFrosch) [inline test/client_tests/client_test.dart:176]: I'm not quite sure if this is the way to go, because the former exception seems to fit better?

File: test/src/client_utils.dart
Comment (LeFrosch) [inline test/src/client_utils.dart:240]: I noticed that the function arguments were never used. But I’m also not sure if this is the right way to go.

File: test/client_tests/client_test.dart
Comment (mraleph) [reply test/client_tests/client_test.dart:176]: This can be fixed by sending `:status` as part of the trailers. gRPC protocol assumes that OnlyTrailers responses contain both trailers and headers. 

I am going to push a commit on top of your work fixing this.

File: test/src/client_utils.dart
Comment (mraleph) [reply test/src/client_utils.dart:240]: Yeah, I think this is good fix.

**Comment by mraleph** [comment]
@LeFrosch okay, so I have spent some time today figuring how the best structure these checks. I think I have arrived to a good approach at the end and pushed some changes into your repo. 

But now I have made so many changes myself, that I should not be the one reviewing this code. I am going to ask @mkustermann to take a look on Monday and then once he approves I will land the code.

**Comment by LeFrosch** [comment]
@mraleph Thanks a lot for the help and effort :)

File: test/grpc_web_test.dart
Comment (mkustermann) [inline test/grpc_web_test.dart:134]: Maybe add a comment explaining that the server will treat `EchoRequest` with `test:` specially by returning errors instead of echoing anything.

File: test/grpc_web_server.dart
Comment (mkustermann) [inline test/grpc_web_server.dart:173]: Are these ignore-all-errors catches necessary here? Could they hide failures in the tests?

File: test/grpc_web_server.dart
Comment (mkustermann) [inline test/grpc_web_server.dart:202]: Could you avoid hard-coding the port?

File: test/client_tests/client_test.dart
Comment (mkustermann) [inline test/client_tests/client_test.dart:298]: I think it cannot be misinterpreted as header frame - http/2 will treat it as data frame(s).

as a header frame -> as a length-prefixed proto message?

(also below)

File: lib/src/shared/streams.dart
Comment (mkustermann) [reply lib/src/shared/streams.dart:135]: There are two kind of responses:

  * `HttpHeaders` + `Data`... + `HttpHeaders` (tailer) + EOS
  * `HttpHeader` EOS

In the first case (normal case) the http `:status 200` is expected in the first `HttpHeaders` and the `grpc-status` is expected in the trailer `HttpHeaders`. 

It seems this code is handling/looking at both. As discussed offline you mentioned the trailer handling looking only at `grpc-status` is in a different place, consider linking to that.

**Review by mkustermann — APPROVED** [review]

File: lib/src/shared/status.dart
Comment (mraleph) [reply lib/src/shared/status.dart:388]: Good point. Specification actually tells us to be lenient and preserve incorrect messages which can't be decoded. Fixed.

File: test/src/client_utils.dart
Comment (mraleph) [reply test/src/client_utils.dart:248]: Done.

File: lib/src/shared/status.dart
Comment (mraleph) [reply lib/src/shared/status.dart:130]: Done.

File: test/grpc_web_server.dart
Comment (mraleph) [reply test/grpc_web_server.dart:173]: Removed `catchError`.

File: test/grpc_web_server.dart
Comment (mraleph) [reply test/grpc_web_server.dart:202]: Good catch. This was left-over from debugging.

File: test/grpc_web_test.dart
Comment (mraleph) [reply test/grpc_web_test.dart:134]: Done.

File: lib/src/shared/streams.dart
Comment (mraleph) [reply lib/src/shared/streams.dart:135]: > If this method throws, which code will catch it and report it to user code?

The `StreamTransformer.bind` implementation (or more specifically `_SinkTransformerStreamSubscription`) takes care of catching exceptions that occur in transformer and forwarding it into the stream as an error. Added a comment explaining this. 

> It seems this code is handling/looking at both. As discussed offline you mentioned the trailer handling looking only at grpc-status is in a different place, consider linking to that.

Comment added.

**Review by mraleph — APPROVED** [review]

**Comment by mraleph** [comment]
Thanks for taking a look @mkustermann. Waiting for tests to go green and then will land.

**Pull request merged**
